agent/util-scripts/pbench-tool-data-sink

```diff
@@ -0,0 +1,551 @@
+        params_str = params_raw.decode("utf-8")
+        # The expected parameters for this "data-sink" is what "channel" to
+        # subscribe to for the tool meister operational life-cycle.  The
+        # data-sink listens for the state transitions, start | stop | send |
+        # terminate, exiting when "terminate" is received, marking the state
+        # in which data is captured.
+        #
+        # E.g. params = '{ "channel": "run-chan",
+        #                  "benchmark_run_dir": "/loo/goo" }'
+        params = json.loads(params_str)
+        channel = params["channel"]
+        benchmark_run_dir = Path(params["benchmark_run_dir"]).resolve()
+    except Exception as ex:
+        logger.error("Unable to fetch and decode parameter key, %s: %s", param_key, ex)
+        return 4
+    else:
+        if not benchmark_run_dir.is_dir():
+            logger.error(
+                "Run directory argument, %s, must be a real" " directory.",
+                benchmark_run_dir,
+            )
+            return 5
+
+    pidfile_name = f"{PROG}.pid"
+    pfctx = pidfile.PIDFile(pidfile_name)
+    with open(f"{PROG}.out", "w") as sofp, open(f"{PROG}.err", "w") as sefp:
+        with daemon.DaemonContext(
+            stdout=sofp,
+            stderr=sefp,
+            working_directory=os.getcwd(),
+            umask=0o022,
+            pidfile=pfctx,
+            files_preserve=[sofp.fileno(), sefp.fileno(), fh.stream.fileno()],
+        ):
+            try:
+                # We have to re-open the connection to the redis server now that we
+                # are "daemonized".
+                try:
+                    redis_server = redis.Redis(host=redis_host, port=redis_port, db=0)
+                except Exception as e:
+                    logger.error(
+                        "Unable to connect to redis server, %s:%s: %s",
+                        redis_host,
+                        redis_port,
+                        e,
+                    )
+                    return 2
+
+                tds_app = ToolDataSink(redis_server, channel, benchmark_run_dir, logger)
+                tds_app.run()
+            except OSError as exc:
+                if exc.errno == errno.EADDRINUSE:
+                    logger.error(
+                        "ERROR - tool data sink failed to start, localhost:8080 already in use"
+                    )
+                else:
+                    logger.exception("ERROR - failed to start the tool data sink")
+            except Exception:
+                logger.exception("ERROR - failed to start the tool data sink")
+            finally:
+                logger.info("Remove pid file ... (%s)", pidfile_name)
+                try:
+                    os.unlink(pidfile_name)
+                except Exception:
+                    logger.exception("Failed to remove pid file %s", pidfile_name)
+
+    return 0
+
+
+if __name__ == "__main__":
+    status = main(sys.argv)
+    sys.exit(status)
```

Comment:
Systemd file would be nice as well

Reply:
I had considered using `systemd`, but that elevates it to a host managed process from a daemon that is run directly from pbench.  It seems better to not send mixed signals that way.